test.xml doesn't report a test case properly when it panic

With a test like this:

```go
func TestPanic(t *testing.T) {
	var buff *bytes.Buffer
	buff.WriteString("hello")
}
```

rules_go would produce a test.xml like this:


```xml
<testsuites>
	<testsuite errors="1" failures="0" skipped="0" tests="1" time="" name="example.com/panic" timestamp="2025-02-21T00:05:35.143Z">
		<testcase classname="panic" name="TestPanic" time="">
			<error message="No pass/skip/fail event found for test" type="">=== RUN   TestPanic&#xA;--- FAIL: TestPanic (0.00s)&#xA;panic: runtime error: invalid memory address or nil pointer dereference [recovered]&#xA;&#x9;panic: runtime error: invalid memory address or nil pointer dereference&#xA;[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x39e09f7]</error>
		</testcase>
	</testsuite>
</testsuites>

```

Due to this issue, the duration of the test case and the test suite are not populated.

---
Some debugging notes:

From the test.log, we can see the the `--- FAIL` event is there:
```
=== RUN   TestPanic
--- FAIL: TestPanic (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x39e09f7]
...
```

Running this test with `go test -json`, we can see there is a `"Action":"fail"`:

```json
{"Time":"2025-02-20T16:11:18.654146-08:00","Action":"start","Package":"uber.com/panic"}
{"Time":"2025-02-20T16:11:18.914593-08:00","Action":"run","Package":"uber.com/panic","Test":"TestPanic"}
{"Time":"2025-02-20T16:11:18.914655-08:00","Action":"output","Package":"uber.com/panic","Test":"TestPanic","Output":"=== RUN   TestPanic\n"}
{"Time":"2025-02-20T16:11:18.914704-08:00","Action":"output","Package":"uber.com/panic","Test":"TestPanic","Output":"--- FAIL: TestPanic (0.00s)\n"}
{"Time":"2025-02-20T16:11:18.917198-08:00","Action":"output","Package":"uber.com/panic","Test":"TestPanic","Output":"panic: runtime error: invalid memory address or nil pointer dereference [recovered]\n"}
{"Time":"2025-02-20T16:11:18.917214-08:00","Action":"output","Package":"uber.com/panic","Test":"TestPanic","Output":"\tpanic: runtime error: invalid memory address or nil pointer dereference\n"}
{"Time":"2025-02-20T16:11:18.91722-08:00","Action":"output","Package":"uber.com/panic","Test":"TestPanic","Output":"[signal SIGSEGV: segmentation violation code=0x2 addr=0x20 pc=0x1008aa80c]\n"}
{"Time":"2025-02-20T16:11:18.917705-08:00","Action":"output","Package":"uber.com/panic","Test":"TestPanic","Output":"exit status 2\n"}
{"Time":"2025-02-20T16:11:18.917726-08:00","Action":"fail","Package":"uber.com/panic","Test":"TestPanic","Elapsed":0}
{"Time":"2025-02-20T16:11:18.917731-08:00","Action":"output","Package":"uber.com/panic","Output":"FAIL\tuber.com/panic\t0.263s\n"}
{"Time":"2025-02-20T16:11:18.91774-08:00","Action":"fail","Package":"uber.com/panic","Elapsed":0.264}
```

However, if I feed the test.log into the [JSON converter](https://github.com/bazel-contrib/rules_go/blob/b1f90178f55de9f9bbef19f596605e0fa0dd1951/go/tools/bzltestutil/test2json.go#L58), the resulting JSON doesn't have the `"Action":"fail"` for `TestPanic`. So it seems the information is lost when converting text to JSON.

--- a/go/tools/bzltestutil/wrap.go
+++ b/go/tools/bzltestutil/wrap.go
@@ -141,6 +141,11 @@ func Wrap(pkg string) error {
 	cmd.Stdout = io.MultiWriter(os.Stdout, streamMerger.OutW)
 	streamMerger.Start()
 	err := cmd.Run()
+	if err != nil {
+		jsonConverter.Write([]byte{marker})
+		jsonConverter.Write(bigFail)
+		jsonConverter.Write([]byte("\n"))
+	}
 	streamMerger.ErrW.Close()
 	streamMerger.OutW.Close()
 	streamMerger.Wait()

--- a/go/tools/bzltestutil/wrap.go
+++ b/go/tools/bzltestutil/wrap.go
@@ -142,6 +142,7 @@ func Wrap(pkg string) error {
 	streamMerger.Start()
 	err := cmd.Run()
 	if err != nil {
+		// force jsonConverter to flush the buffer, so we get the "fail" event when a test case panics.
 		jsonConverter.Write([]byte{marker})
 		jsonConverter.Write(bigFail)
 		jsonConverter.Write([]byte("\n"))

--- a/go/tools/bzltestutil/wrap.go
+++ b/go/tools/bzltestutil/wrap.go
@@ -141,15 +141,15 @@ func Wrap(pkg string) error {
 	cmd.Stdout = io.MultiWriter(os.Stdout, streamMerger.OutW)
 	streamMerger.Start()
 	err := cmd.Run()
+	streamMerger.ErrW.Close()
+	streamMerger.OutW.Close()
+	streamMerger.Wait()
 	if err != nil {
 		// force jsonConverter to flush the buffer, so we get the "fail" event when a test case panics.
 		jsonConverter.Write([]byte{marker})
 		jsonConverter.Write(bigFail)
 		jsonConverter.Write([]byte("\n"))
 	}
-	streamMerger.ErrW.Close()
-	streamMerger.OutW.Close()
-	streamMerger.Wait()
 	jsonConverter.Close()
 	if out, ok := os.LookupEnv("XML_OUTPUT_FILE"); ok {
 		werr := writeReport(jsonBuffer, pkg, out)

--- a/tests/core/go_test/BUILD.bazel
+++ b/tests/core/go_test/BUILD.bazel
@@ -285,6 +285,11 @@ go_bazel_test(
     srcs = ["timeout_test.go"],
 )
 
+go_bazel_test(
+    name = "xml_panic_test",
+    srcs = ["xml_panic_test.go"],
+)
+
 # Tests using .syso files in go_binary both transitively and directly.
 go_test(
     name = "syso_transitive_test",

--- a/tests/core/go_test/xml_panic_test.go
+++ b/tests/core/go_test/xml_panic_test.go
@@ -46,7 +46,9 @@ type xmlTestCase struct {
 }
 type xmlTestSuite struct {
 	XMLName   xml.Name      `xml:"testsuite"`
+	Name    string     `xml:"name,attr"`
 	TestCases []xmlTestCase `xml:"testcase"`
+	Time      string         `xml:"time,attr"`
 }
 type xmlTestSuites struct {
 	XMLName xml.Name       `xml:"testsuites"`
@@ -80,6 +82,9 @@ func Test(t *testing.T) {
 	}
 
 	for _, suite := range suites.Suites {
+		if suite.Time == "" {
+			t.Errorf("empty time attribute for test suite %q", suite.Name)
+		}
 		for _, tc := range suite.TestCases {
 			if tc.Time == "" {
 				t.Errorf("empty time attribute for test case %q", tc.Name)

--- a/tests/core/go_test/xml_panic_test.go
+++ b/tests/core/go_test/xml_panic_test.go
@@ -1,0 +1,92 @@
+package xml_panic_test
+
+import (
+	"encoding/xml"
+	"errors"
+	"io/ioutil"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/bazelbuild/rules_go/go/tools/bazel_testing"
+)
+
+func TestMain(m *testing.M) {
+	bazel_testing.TestMain(m, bazel_testing.Args{
+		Main: `
+-- BUILD.bazel --
+load("@io_bazel_rules_go//go:def.bzl", "go_test")
+
+go_test(
+	name = "panic_test",
+	srcs = ["panic_test.go"],
+)
+-- panic_test.go --
+package panic_test
+
+import "testing"
+
+func TestPanic(t *testing.T) {
+	panic("panic by design")
+}
+`,
+	})
+}
+
+type testError struct {
+	Message string `xml:"message,attr"`
+}
+
+type xmlTestCase struct {
+	XMLName xml.Name   `xml:"testcase"`
+	Name    string     `xml:"name,attr"`
+	Time    string     `xml:"time,attr"`
+	Error   *testError `xml:"error"`
+}
+type xmlTestSuite struct {
+	XMLName   xml.Name      `xml:"testsuite"`
+	TestCases []xmlTestCase `xml:"testcase"`
+}
+type xmlTestSuites struct {
+	XMLName xml.Name       `xml:"testsuites"`
+	Suites  []xmlTestSuite `xml:"testsuite"`
+}
+
+func Test(t *testing.T) {
+	err := bazel_testing.RunBazel("test", "//:panic_test", "--test_env=GO_TEST_WRAP_TESTV=1")
+	if err == nil {
+		t.Fatal("expected test to fail")
+	}
+	var exErr *exec.ExitError
+	if !errors.As(err, &exErr) {
+		t.Fatalf("unexpected error: %v", err)
+	} else if exErr.ExitCode() != 3 {
+		t.Fatalf("unexpected exit code %d\n%s", exErr.ExitCode(), exErr.Stderr)
+	}
+	p, err := bazel_testing.BazelOutput("info", "bazel-testlogs")
+	if err != nil {
+		t.Fatalf("could not find testlog root: %s", err)
+	}
+	path := filepath.Join(strings.TrimSpace(string(p)), "panic_test/test.xml")
+	b, err := ioutil.ReadFile(path)
+	if err != nil {
+		t.Fatalf("could not read generated xml file: %s", err)
+	}
+
+	var suites xmlTestSuites
+	if err := xml.Unmarshal(b, &suites); err != nil {
+		t.Fatalf("could not unmarshall generated xml: %s", err)
+	}
+
+	for _, suite := range suites.Suites {
+		for _, tc := range suite.TestCases {
+			if tc.Time == "" {
+				t.Errorf("empty time attribute for test case %q", tc.Name)
+			}
+			if tc.Error != nil {
+				t.Errorf("unexpected error for test case %q: %s", tc.Name, tc.Error.Message)
+			}
+		}
+	}
+}

